pkg/attribute/bag.go

```diff
@@ -41,6 +41,12 @@ type Bag interface {
 	// Bytes returns the named attribute if it exists.
 	Bytes(name string) ([]uint8, bool)
 
+	// Foreach allows a caller to process all of the (key, value) pairs this bag knows about. If the provided func
+	// returns false iteration is stopped.
+	//
+	// The view into the values is provided in this form so that the bag can keep iteration thread safe.
+	ForEach(handler func(key string, val interface{}) bool)
```

Comment:
I forgot about the reentry problem. The tracing code will only ever need the keys that correspond to string attributes. I guess the question becomes: do we care about providing a handle to let a user iterate over all keys, do we want to provide a handle to iterate over the keys of each attribute type (StringKeys, Int64Keys, etc) so we get more granular locks and less allocations in the common case, or do we want to provide a way to get only the keys we need right now (the string keys). Thoughts?